zio-http/src/main/scala/zio/http/model/headers/values/Authorization.scala

```diff
@@ -0,0 +1,147 @@
+      nc: Int,
+      userhash: Boolean,
+      rest: Map[String, String],
+    ) extends AuthScheme
+    final case class Bearer(token: String)                                extends AuthScheme
+    final case class Unparsed(authScheme: String, authParameters: String) extends AuthScheme
+  }
+
+  /**
+   * The Authorization header value contains one of the auth schemes
+   * [[AuthScheme]].
+   */
+  final case class AuthorizationValue(authScheme: AuthScheme) extends Authorization
+
+  /** The Authorization header value is invalid. */
+  case object InvalidAuthorizationValue extends Authorization
+
+  def fromAuthorization(header: Authorization): String = header match {
+    case AuthorizationValue(Basic(username, password)) =>
+      s"Basic ${Base64.getEncoder.encodeToString(s"$username:$password".getBytes(StandardCharsets.UTF_8))}"
+
+    case AuthorizationValue(Digest(response, username, realm, uri, opaque, algo, qop, cnonce, nc, userhash, rest)) =>
+      val restOfParams = {
+        val asStrings = rest.map(e => s"""${e._1}="${e._2}"""")
+        if (asStrings.isEmpty) "" else asStrings.mkString(",", ",", "")
+      }
+
+      s"""$Digest response="$response",username="$username",realm="$realm",uri=${uri.toString},opaque="$opaque",algorithm=$algo,""" +
+        s"""qop=$qop,cnonce="$cnonce",nc=$nc,userhash=${userhash.toString}$restOfParams"""
+    case AuthorizationValue(Bearer(token))            => s"$Bearer $token"
+    case AuthorizationValue(Unparsed(scheme, params)) => s"$scheme $params"
+    case InvalidAuthorizationValue                    => ""
+  }
+
+  private final val ExpectedParams =
+    Set("response", "username", "username*", "realm", "uri", "opaque", "algorithm", "qop", "cnonce", "nc", "userhash")
+  private final val quotationMark  = "\""
+  private final val comma          = ","
+
+  private sealed trait State
+  private object State {
+    object Key   extends State
+    object Value extends State
+  }
+
+  def toAuthorization(value: String): Authorization = {
+    val parts = value.split(' ')
+    if (parts.length >= 2) {
+      parts(0).toLowerCase match {
+        case "basic"  =>
+          val partsOfBasic = new String(Base64.getDecoder.decode(parts(1))).split(':')
+          if (partsOfBasic.length == 2) AuthorizationValue(Basic(partsOfBasic(0), partsOfBasic(1)))
+          else InvalidAuthorizationValue
+        case "digest" => parseDigest(parts.tail.mkString(" "))
+        case "bearer" => AuthorizationValue(Bearer(parts(1)))
+        case _        => AuthorizationValue(Unparsed(parts(0), parts.tail.mkString(" ")))
+      }
+    } else InvalidAuthorizationValue
+  }
+
+  // https://datatracker.ietf.org/doc/html/rfc7616
+  private def parseDigest(value: String) = try {
+    var state: State         = State.Key
+    var index                = 0
+    var params               = HashMap[String, String]()
+    var rest                 = HashMap[String, String]()
+    var currentKey: String   = null
```

Comment:
Is there a way to get rid of those `var` and `nulls`?

Reply:
Sure there is. However I was going for low-level implementation to avoid unnecessary allocations. If you think it's not necessary here I can reimplement it differently.

Reply:
So how about now?